### PR TITLE
#1868 Use MANE transcript in variant title

### PR DESCRIPTION
**Pull request by shaungc-su**
Github issue containing the requirement: #1868 

Test steps used by curators are in this doc: 
https://slack-files.com/T0F6S42DA-FSLNF16TU-ca10f0a037

~~Special notes for this PR:~~ (synced)
- ~~I excluded all the changes we borrowed from functional data branch to be prepared for our pull request. This may result in the travis test failed, which will be fixed once functional data's PR (#2013) is merged and synced with this PR. So travis failed build can be ignored for now, just to point it out.~~
- ~~If you are considering local dev test, this PR will not be able to be `git checkout` and run for now since I excluded the functional data change. You can still test it using the test instance, which contains functional data change for the `/ldh/<CA ID>` endpoint.~~
- ~~This pull request cannot be merged into dev until functional data's PR is merged and synced with this PR.~~

Notes related to development:
- Adds JavaScript 2016,2017 (ES7,ES8) support to use async/await, so that MANE's fetch logic can be better integrated with existing code. To run locally on this branch, or after merged into dev, have to reinstall the frontend dependency, or run `make clean  && python3.4 bootstrap.py -v 2.9.5 --setuptools-version 15.2 && bin/buildout -c buildout-dev.cfg`.

### CODE REVIEW

File: src/clincoded/static/components/add_external_resource.js [hunk 1]
```diff
@@ -735,32 +751,42 @@ function carQueryResource() {
...
+                        // something failed with the parsing of ClinVar data; roll back to CAR data (just use initial value of `finalState`)
                     }
-                }).catch(e => {
+                }
+                catch (e) {
                     // error handling for ClinVar query
                     this.setFormErrors('resourceId', 'Error querying ClinVar for additional data. Please check your input and try again.');
-                    this.setState({queryResourceBusy: false, resourceFetched: false});
-                });
+                    finalState.resourceFetched = false;
```
Comment (bryanwulf) [inline src/clincoded/static/components/add_external_resource.js]: To maintain previous `setState` logic, I'd recommend adding a removal of `tempResource` (from `finalState`) here.

File: src/clincoded/static/components/add_external_resource.js [hunk 2]
```diff
@@ -785,29 +811,43 @@ function carQueryResource() {
...
                 } else {
                     // Fall back to CAR data without the canonical transcript title if parsing fails
                     if (data['tempAlleles']) delete data['tempAlleles'];
-                    this.setState({queryResourceBusy: false, tempResource: data, resourceFetched: true});
                 }
             } else {
                 // in case the above two fail (theoretically a 404 json response, but an error is thrown instead (see below))
                 this.setFormErrors('resourceId', 'CA ID not found');
-                this.setState({queryResourceBusy: false, resourceFetched: false});
+                finalState.resourceFetched = false;
```
Comment (bryanwulf) [inline src/clincoded/static/components/add_external_resource.js]: To maintain previous `setState` logic, I'd recommend adding a removal of `tempResource` (from `finalState`) here.

File: src/clincoded/static/components/add_external_resource.js [hunk 3]
```diff
@@ -894,3 +937,41 @@ function carSubmitResource(func) {
         });
     }
 }
+
+
+async function queryManeTranscriptTitle(carId, carJson) {
```
Comment (bryanwulf) [inline src/clincoded/static/components/add_external_resource.js]: Please add a function declaration comment block for `queryManeTranscriptTitle`.

File: src/clincoded/static/components/variant_curation.js [hunk 4]
```diff
@@ -606,7 +606,7 @@ class VariantAddHistory extends Component {
 
         return (
             <div>
-                <span>Variant <strong><a href={"/variant-central/?variant=" + variant.uuid}>{variant.clinvarVariantTitle ? variant.clinvarVariantTitle : (variant.hgvsNames.GRCh38 ? variant.hgvsNames.GRCh38 : variant.hgvsNames.GRCh37)}</a></strong> added</span>
+                <span>Variant <strong><a href={"/variant-central/?variant=" + variant.uuid}>{renderVariantTitle(variant)}</a></strong> added</span>
```
Comment (bryanwulf) [inline src/clincoded/static/components/variant_curation.js]: It doesn't appear that `renderVariantTitle` has been defined/declared in this file.

File: src/clincoded/static/libs/get_mane_transcript.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ * @param {object} genomicCarJson - Genomic CAR json response object.
+ * @returns {(string|null)} MANE transcript id.
+ */
+export const parseManeTranscriptIdFromGenomicCar = (genomicCarJson) => {
+    const {
+        externalRecords: {
+            MANEPrefRefSeq: {
+                id = null
+            } = {}
+        }
```
Comment (bryanwulf) [inline src/clincoded/static/libs/get_mane_transcript.js]: Unless there's a reason **not** to do so, I would suggest setting a default value (of an empty object).

File: src/clincoded/static/libs/render_variant_title.js [hunk 6]
```diff
@@ -14,6 +14,8 @@ export function renderVariantTitle(variant, stringOnly) {
     if (variant) {
         if (variant.clinvarVariantId && variant.clinvarVariantTitle) {
             variantTitle = variant.clinvarVariantTitle;
+        } else if (variant.maneTranscriptTitle) {
```
Comment (bryanwulf) [inline src/clincoded/static/libs/render_variant_title.js]: It wouldn't hurt to update the function declaration comment block (to include this new option).

File: src/clincoded/static/libs/get_mane_transcript.js (hunk 5 above)
Comment (shaungc-su) [reply src/clincoded/static/libs/get_mane_transcript.js]: `id` was default to `null`, changing default to an empty string.

File: src/clincoded/static/components/variant_curation.js (hunk 4 above)
Comment (shaungc-su) [reply src/clincoded/static/components/variant_curation.js]: Good catch! Going to fix this.

**Comment by shaungc-su** [comment]
Fixed changes based on Bryan's feedback. Will now work on async/await -> Promise conversion on top of that.

File: src/clincoded/static/libs/get_mane_transcript.js (hunk 5 above)
Comment (shaungc-su) [reply src/clincoded/static/libs/get_mane_transcript.js]: Adding the missing default value for `externalRecords`, nice catch!

**Comment by shaungc-su** [comment]
Closing the PR in favor of the updated PR: https://github.com/ClinGen/clincoded/pull/2159

**Pull request closed without merging**